autoload/neomake.vim

```diff
@@ -525,41 +531,41 @@ function! neomake#MakeHandler(job_id, data, event_type) abort
     let jobinfo = s:jobs[a:job_id]
     let maker = jobinfo.maker
     if index(['stdout', 'stderr'], a:event_type) >= 0
-        let lines = a:data
-        if has_key(maker, 'mapexpr')
-            let lines = map(copy(lines), maker.mapexpr)
-        endif
-
-        for line in lines
-            call neomake#utils#DebugMessage(
-                \ get(maker, 'name', 'makeprg').' '.a:event_type.': '.line)
-        endfor
+        call neomake#utils#DebugMessage(
+            \ get(maker, 'name', 'makeprg').' '.a:event_type.': ["'.join(a:data, '", "').'"]')
         call neomake#utils#DebugMessage(
             \ get(maker, 'name', 'makeprg').' '.a:event_type.' done.')
 
         " Register job output. Buffer registering of output for long running
         " jobs.
         let last_event_type = get(jobinfo, 'event_type', a:event_type)
         let jobinfo.event_type = a:event_type
+
+        " a:data is a List of 'lines' read. Each element *after* the first
+        " element represents a newline
         if has_key(jobinfo, 'lines')
             " As per https://github.com/neovim/neovim/issues/3555
             let jobinfo.lines = jobinfo.lines[:-2]
-                        \ + [jobinfo.lines[-1] . get(lines, 0, '')]
-                        \ + lines[1:]
+                        \ + [jobinfo.lines[-1] . get(a:data, 0, '')]
+                        \ + a:data[1:]
         else
-            let jobinfo.lines = lines
+            let jobinfo.lines = a:data
         endif
+
         let now = localtime()
         if (!maker.buffer_output || last_event_type !=# a:event_type) ||
                 \ (last_event_type !=# a:event_type ||
                 \  now - jobinfo.start < 1 ||
                 \  now - jobinfo.last_register > 3)
-            call s:RegisterJobOutput(jobinfo, maker, jobinfo.lines)
-            unlet jobinfo.lines
+            call s:RegisterJobOutput(jobinfo, maker, jobinfo.lines[:-2])
+            let jobinfo.lines = jobinfo.lines[-1:]
             let jobinfo.last_register = now
         endif
     elseif a:event_type ==# 'exit'
         if has_key(jobinfo, 'lines')
+            if jobinfo.lines[-1] == ''
+                call remove(jobinfo.lines, -1)
+            endif
             call s:RegisterJobOutput(jobinfo, maker, jobinfo.lines)
         endif
```

Comment:
@nerdrew 
I think this is wrong and there should be no call to `RegisterJobOutput` at all on "exit": exit is not meant to receive new data.
Also the change/slicing to lines above feels weird.